Sources/SwiftProtobuf/Message+AsyncMessageSequence.swift

```diff
@@ -0,0 +1,232 @@
+  ///    messages encoded by this sequence, or in messages nested within these messages.
+  ///   - partial: If `false` (the default),  after decoding a message, `Message.isInitialized`
+  ///     will be checked to ensure all fields are present. If any are missing,
+  ///     `BinaryEncodingError.missingRequiredFields` will be thrown.
+  ///   - options: The BinaryDecodingOptions to use.
+  /// - Returns: An asynchronous sequence of messages read from the `AsyncSequence` of bytes.
+  /// - Throws: `BinaryDecodingError` if decoding fails, throws
+  ///           `BinaryDelimited.Error` for some reading errors,
+  ///           `BinaryDecodingError.truncated` if the sequence ends before fully decoding a
+  ///           message or a delimiter,
+  ///           `BinaryDecodingError.malformedProtobuf`if a delimiter could not be read and
+  ///           `BinaryDecodingError.tooLarge` if a size delimiter of 2GB or greater is found.
+  ///
+public static func asyncSequence<Base: AsyncSequence> (
+    baseSequence: Base,
+    extensions: ExtensionMap? = nil,
+    partial: Bool = false,
+    options: BinaryDecodingOptions = BinaryDecodingOptions()
+  ) -> AsyncMessageSequence<Base, Self> {
+    AsyncMessageSequence<Base, Self>(
+      baseSequence: baseSequence,
+      extensions: extensions,
+      partial: partial,
+      options: options
+    )
+  }
+}
+
+@available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+extension AsyncSequence {
+  
+  /// Creates an asynchronous sequence of size-delimited messages from this sequence of bytes.
+  /// Delimited format allows a single file or stream to contain multiple messages. A delimited message
+  /// is a varint encoding the message size followed by a message of exactly that size.
+  ///
+  /// - Parameters:
+  ///   - messageType: The  type of message to read.
+  ///   - extensions: An `ExtensionMap` used to look up and decode any extensions in
+  ///    messages encoded by this sequence, or in messages nested within these messages.
+  ///   - partial: If `false` (the default),  after decoding a message, `Message.isInitialized`
+  ///     will be checked to ensure all fields are present. If any are missing,
+  ///     `BinaryEncodingError.missingRequiredFields` will be thrown.
+  ///   - options: The BinaryDecodingOptions to use.
+  /// - Returns: An asynchronous sequence of messages read from the `AsyncSequence` of bytes.
+  /// - Throws: `BinaryDecodingError` if decoding fails, throws
+  ///           `BinaryDelimited.Error` for some reading errors,
+  ///           `BinaryDecodingError.truncated` if the stream ends before fully decoding a
+  ///           message or a delimiter,
+  ///           `BinaryDecodingError.malformedProtobuf`if a delimiter could not be read and
+  ///           `BinaryDecodingError.tooLarge` if a size delimiter of 2GB or greater is found.
+  public func decodedBinaryDelimitedMessages<M: Message>(
```

Comment:
@thomasvl @glbrntt Do you think that name makes sense? I think it does but just want to get a few more opinions on it.

```suggestion
  public func decodedBinaryDelimitedMessages<M: Message>(
```

Reply:
The only nit I can see is we usually say _parse_ not _decode_ on the public apis, so maybe `parseBinaryDelimitedMessages`?  But both names also fee wrong since it sounds a little to _active_ the api is created a new sequence to iterate on, but I'm not sure what would be the best naming for that.

Reply:
`binaryDelimitedMessages` works for me!

Reply:
You're probably right it's parsing not decoding we're doing here. 

Both the [C#](https://cloud.google.com/dotnet/docs/reference/Google.Protobuf/latest/Google.Protobuf.MessageParser
) and [Java](https://github.com/protocolbuffers/protobuf/blob/023b43dacf7b0a1f6ded403a46965d277e8cbf3c/java/core/src/main/java/com/google/protobuf/Parser.java#L225) APIs use `parseDelimitedFrom(Stream)`

But that doesn't make sense here because the extension is defined on the byte sequence, not the message.

maybe
`bytes.parseDelimitedSequence(of: MessageType.self)
`
or
`bytes.parseDelimitedMessages(of: MessageType.self)
`
?

There is a [discussion](https://github.com/protocolbuffers/protobuf/issues/10229) in the protobuf repo about harmonizing binary delimited APIs.

Reply:
These are Swift APIs though, so we should aim to be harmonious with [Swift's API design guidelines](https://www.swift.org/documentation/api-design-guidelines/) around naming, where in general value-returning functions return noun phrases describing the value being returned. `parse...` is an active verb phrase. (The existing `BinaryDelimited` functions in SwiftProtobuf don't obey those, unfortunately, but we should also revisit _that_ in the future.)

So a more Swift-appropriate name that better matches the C#/Java terms would be `parsedDelimitedMessages`, which reads oddly to me. I also don't think we should prefer matching other languages' naming here; the goal of protobuf is to provide—as much as one can—natural interfaces in the target language for the generated messages.

`binaryDelimitedMessages` seems like a nice compromise around not being too verbose, explaining what the function does, and fitting well with other async stream methods/properties that use the noun-phrase naming rule.

Reply:
Is binary even necessary given the context? Protobuf is primarily a binary serialization format.

`sizeDelimitedMessages` would fit the guidelines and describe the format more specifically. The format is described as [length delimited](https://protobuf.dev/programming-guides/encoding/#length-types) in the protobuf docs but size seems more appropriate for Swift.

Reply:
Sticking with the protobuf terminology is generally appropriate:  We're optimizing for people who either see this in SwiftProtobuf and are searching for more information or know it from some other language and are trying to use it in SwiftProtobuf.

Reply:
> Is binary even necessary given the context? Protobuf is primarily a binary serialization format.

Protobuf (and the Swift impl included), have three possible formats: serialized binary, JSON, and TextFormat. When you look at extensions each format adds to `Message`, I think there's benefit it always including the format in the name - it helps auto complete, consistency between apis using formats, etc.

I fear someone that has seen _JSON Streaming_ apis might think `parseDelimitedMessages` follows that _standard_ of a newline between messages, and a JSON payload on each line. So calling out the _serialized_ or _binary_ for protobuf naming might be good to steer folks expectations a little better.

Having said, that, _naming_ is always one of the classic programming problems, so it comes down to balance between all the competing things.

Reply:
Right, I'm not suggesting we should unnecessarily diverge from protobuf terminology, but we shouldn't do it at the expense of Swift naming conventions either. It's definitely possible to marry the two.

I'm fine with `sizeDelimitedMessages`, if nobody else objects.

Reply:
And what would we use for a hypothetical streaming JSON decoder?  (Which I can imagine us wanting to implement someday.)

Reply:
Ok thanks for the input everyone, I think I'm going to go with `binaryDelimitedMessages` as that seems the least controversial for now.